### ISSUE TITLE
Accessing user information AND authorization code

### ISSUE THREAD

**Issue by serkolch** [google/google-api-javascript-client#328]
Hi there,

I have a BE already set up for Google Sign In (works on our Android app) that requires user information and an authorization code to validate. My question is two-fold:
1- Is it possible to access the authorization code from the authentication behavior?
2- Can I get user information from the authorization flow?

In scenario 1 (authentication), I am able to pull a googleUser and get their basic information and their access_token and id_token with getAuthResponse(true), but am unable to see authorization code anywhere in the response. In this case, all I really do is initialize and attachClickHandler to my React element:

initialize

	gapi.load('auth2', () => {
		gapi.auth2.init({
			client_id: `${googleSignInID}.apps.googleusercontent.com`,
		})
	});

React element

	componentDidMount () {
		this.attachClickHandler(
			this.button, // React ref
			{}, // options
			this.onGoogleSuccess,
			this.onGoogleFailure,
		);
	}

this.onGoogleSuccess has access to the googleUser response as expected.

---

In scenario 2 (authorization), trying to initialize as above returns an error (as expected based on warnings from the documentation). However, just loading the client gives a response with user information:

initialize

	gapi.load('auth2', () => {});

React element

	onGoogleClick () {
		gapi.auth2.authorize({
			client_id: `${googleSignInID}.apps.googleusercontent.com`,
			scope: 'email profile openid code',
			response_type: 'id_token permission code'
		}, response => {
			if (response.error) {
				// An error happened!
				return;
			}
			debugger;
		});
	}

The response here seems to be a subset of the response from behavior 1 (although this includes the authorization code).

---

To recap:
- Can I get authorization code from the authentication flow?
- Can I get user information from the authorization flow?
- Is there a way to fix one or combine the two to get the full information?

Thanks

**Comment by TMSCH** [comment]
Hi @serkolch, the user's information is actually contained in the `id_token` returned. When using `gapi.auth2.init`/`signIn`, which has a notion of "User", the library unpacks it and creates the `GoogleUser` object with basic profile information.
If you want to get both the `code` and the profile information, you'll have to use scenario 2 and decode the `id_token` by using a JWT library decoder.

**Comment by serkolch** [comment]
@TMSCH 
Thank you so much! I was struggling too long with this before I opened a stack overflow thread and an issue here. I don't know if I'm just stupid / tired or if the documentation is lacking, but it just did not click to me that the id_token could be decoded.

**Comment by TMSCH** [comment]
@serkolch glad that solves your issue! There is indeed no mention in the Google Sign-In JS SDK doc on decoding the ID Token client side (in JavaScript), as it is usually unnecessary, so it's not your fault! It is mentioned in the [backend auth section](https://developers.google.com/identity/sign-in/web/backend-auth) though.

**Comment by mesqueeb** [comment]
@TMSCH Do you have any low-footprint recommendations for decoding the ID token with Javascript?

**Comment by mesqueeb** [comment]
Just wanted to leave this for others.
This is the JS only method to decode the `id_token`:
```js
function parseJwt (token) {
  let base64Url = token.split('.')[1]
  let base64 = base64Url.replace('-', '+').replace('_', '/')
  return JSON.parse(window.atob(base64))
}
```

**Comment by leonasa** [comment]
@mesqueeb did you get it? were you able to use combine with gapi.auth2.authorize  and firebaase to authorise calendar access.
what about the push events?

**Comment by mesqueeb** [comment]
@leonasa what push events?

**Comment by leonasa** [comment]
@mesqueeb, thank you for you reply.
I am trying to combine "gapi.auth2.authorize and firebaase to authorise calendar"
But I need an offline/server authentication also. So, I am gapi.auth2 to get an Auth Code to be used in the backend. and be able to get the updates on [https://developers.google.com/calendar/v3/push](url)
For now, I can get it to work.
But I am not able to authenticate to Firebase...
